### PR TITLE
bpo-42128: Structural Pattern Matching (PEP 634)

### PR DESCRIPTION
**Pull request by brandtbucher**
PEP 634 has now been accepted, and we'd like to get this into the next alpha.

Several people have volunteered to review the implementation, since it's so huge. Other reviews are very welcome, if anybody has a bit of time to pitch in. This touches tons of stuff: the parser, the compiler, the VM, the builtins, the stdlib, the tests... I'd like as many eyeballs as possible!

https://bugs.python.org/issue42128

### CODE REVIEW

File: Lib/ast.py
Comment (brandtbucher) [inline Lib/ast.py]: I'm not 100% sure how the operator precedence machinery works here, but this feels correct to me and behaves as expected. I think `TEST`, `EXPR`, and `ATOM` are quite arbitrary, but the order is correct, so whatever. 🙂

File: Lib/opcode.py
Comment (brandtbucher) [inline Lib/opcode.py:67]: This diff looks worse than it really is.

Some opcodes were out-of-order or had incorrect vertical whitespace, which meant that the new opcodes actually shadowed / were shadowed by other opcodes when I first added them. It was confusing (to say the least), so I fixed it.

File: Lib/test/test_patma.py
Comment (brandtbucher) [inline Lib/test/test_patma.py:2877]: Copy and paste this command to run a small benchmark suite generated from our tests.

File: Parser/Python.asdl
Comment (brandtbucher) [inline Parser/Python.asdl:94]: I added this new node since our review Tuesday. We used to use `BinOp` here.

File: Python/ceval.c
Comment (brandtbucher) [inline Python/ceval.c]: This file isn't ready for review yet (still have some cleanup / TODOs)...

File: Python/compile.c
Comment (brandtbucher) [inline Python/compile.c:218]: One other change I made here: `compiler_error` takes varargs now, like `compiler_warn`. I got tired of building strings for helpful error messages. 🙃

File: Python/compile.c
Comment (brandtbucher) [inline Python/compile.c]: This comes in handy to keep visual clutter down in the nastier compound patterns.

File: Python/compile.c
Comment (brandtbucher) [inline Python/compile.c]: NB!

Also, note that the current pattern compiler is designed to be as simple and easy-to-review as possible. The PEP gives us broad freedom to do trickier stuff, but I'm not going to begin work on a rewrite until we have a couple of releases using this simpler approach.

File: Lib/ast.py
Comment (isidentical) [reply Lib/ast.py]: For the `MatchOr`, it could only require the parentheses on `_Precedence.BOR` (same with `EXPR`, but more explicit) (I would also go for changing the `EXPR` in the `MatchAs` with the `BOR`). For the patterns inside of binary-or, we could simply mark them with the next precedence that comes right after the `BOR`, which is `BXOR` (so that we could avoid adding parentheses around the complex numbers).

Comment (brandtbucher) [reply Lib/ast.py]: How's something like this?

```suggestion
    def visit_MatchAs(self, node):
        with self.require_parens(_Precedence.TEST, node):
            self.set_precedence(_Precedence.BOR, node.pattern)
            self.traverse(node.pattern)
            self.write(f" as {node.name}")

    def visit_MatchOr(self, node):
        with self.require_parens(_Precedence.BOR, node):
            self.set_precedence(_Precedence.BOR.next(), *node.patterns)
            self.interleave(lambda: self.write(" | "), self.traverse, node.patterns)

```

I think I like `_Precedence.BOR.next()` better than `_Precedence.BXOR`.

Comment (brandtbucher) [reply Lib/ast.py]: ...well that's a stupid diff.

Comment (isidentical) [reply Lib/ast.py]: > I think I like _Precedence.BOR.next() better than _Precedence.BXOR.

That's clever! (diff LGTM)

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: Okay, done. Ready for review.

**Comment by brandtbucher** [comment]
So I see three categories of buildbot failures:

- Memory leaks in test_io (all Refleak jobs).
- Undefined references to _PyMangle and PyAST_CompileObject during LTO (all Fedora Stable LTO jobs).
- Segfaults in testlib2to3 (PPC64LE RHEL7 LTO, s390x Fedora LTO PR, s390x RHEL8 LTO + PGO).

I’ve verified that all of these same failures occur on the 3.x buildbots, so I think they can be safely ignored.

**Comment by pablogsal** [comment]
> So I see three categories of buildbot failures:

I am going to try to fix the segfaults first, as we have a release soon of 3.10 alpha 2 and I don't want those segfaults to be there.

File: Python/ast_opt.c
Comment (ncoghlan) [inline Python/ast_opt.c:825]: Something I noticed while working on PEP 642's reference implementation is that this ast_opt code is currently relying on the parser to keep the operators that it doesn't handle away from it, so it's easy to segfault by feeding it your own AST tree, without invoking the parser.

While filling in `validate_pattern` in ast.c will be the main answer to that, I think there are some genuinely missing cases here as well (e.g. `+1` will fail the UnaryOp check).

Comment (isidentical) [reply Python/ast_opt.c:825]: > While filling in validate_pattern in ast.c will be the main answer to that, I think there are some genuinely missing cases here as well (e.g. +1 will fail the UnaryOp check).

The [drafted AST validator](https://github.com/isidentical/cpython/tree/patma-ast-validators) should prevent `+1`, but I can't say for sure is whether a full coverage exists (I tried to implement it by looking at the grammar but it would be nice if @brandtbucher can do a second pass on it): 
```
>>> t = ast.parse("match test:\n\tcase -1: pass")
>>> t.body[0].cases[0].pattern.op = ast.UAdd()
>>> print(ast.unparse(t))
match test:
    case +1:
        pass
>>> compile(t, "<test>", "exec")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: only USub operator is allowed for UnaryOp nodes inside of a pattern
```

Comment (brandtbucher) [reply Python/ast_opt.c:825]: > Something I noticed while working on PEP 642's reference implementation is that this ast_opt code is currently relying on the parser to keep the operators that it doesn't handle away from it, so it's easy to segfault by feeding it your own AST tree, without invoking the parser.

I'm not sure it's easy to actually segfault the compiler with malformed patterns (at least `--with-pydebug`); most of the assumptions it makes here and in `compile.c` about what the parser can emit are guarded either by `assert(...)` or `Py_UNREACHABLE()` calls. (I'm probably a heavier user of `assert` than the historical authors of these files). As I see it, the goal of `PyAST_Validate` is to make sure these checks cannot be hit by people playing with `ast` and `compile` in userland.

So checking those same assumptions in `PyAST_Validate` should be sufficient (and more should be added if missing).

> While filling in `validate_pattern` in ast.c will be the main answer to that, I think there are some genuinely missing cases here as well (e.g. `+1` will fail the UnaryOp check).

For what it's worth, `+1` isn't valid. Full grammar [here](https://www.python.org/dev/peps/pep-0634/#appendix-a-full-grammar).

> I can't say for sure is whether a full coverage exists (I tried to implement it by looking at the grammar but it would be nice if @brandtbucher can do a second pass on it)

Ha! I knew I was forgetting something. I promised you a review, then promptly got distracted with `ceval.c` and forgot. Thanks for reminding me.

**Comment by brandtbucher** [comment]
Also, if any core devs are eavesdropping and haven't voted for Batuhan's promotion yet, please [do so](https://discuss.python.org/t/vote-to-promote-batuhan-taskaya)! He's absolutely earned it.

File: Python/ast_opt.c
Comment (brandtbucher) [reply Python/ast_opt.c:825]: @isidentical can you open up a PR in your own repo (merging `isidentical/cpython:patma-ast-validators` into `isidentical/cpython:master`)? That way I can leave review comments.

Comment (isidentical) [reply Python/ast_opt.c:825]: isidentical/cpython#1 should work

Comment (brandtbucher) [reply Python/ast_opt.c:825]: Great, thanks!

**Comment by markshannon** [comment]
```python
def f(t):
    a, b, p, q, x, y = [None]*6
    match t:
        case []:
            pass
        case [x]:
            y = 2
        case [a, b] if a == b:
            pass
        case [p, q]:
            pass
    print(a,b,p,q,x,y)

f((1,2))
```
This prints `1 2 1 2 None None`, instead of `None None 1 2 None None`.
Since `[a, b] if a == b` doesn't match, why are values assigned to `a` and `b`?
`[x]` doesn't match, and `x` is not assigned to.

**Comment by brandtbucher** [comment]
Hi Mark.

PEP 634 says (in regard to name bindings in patterns):

> These may happen earlier or later depending on the implementation strategy, the only constraint being that capture variables must be set before guards that use them explicitly are evaluated.

In general, there are two points to keep in mind for cases like yours:

- Cases consist of patterns and guards. These are separate conditions. While the language of the PEP allows checking guard conditions while matching the pattern (which this implementation doesn't do), the constraint on that is that any names used by the guard have already been bound, as you experienced.

- The implementation, while still quite naïve at this point, will quit matching as soon as a pattern is proven to be a non-match. In your first and second cases, we jump to the next case as soon as we observe a sequence of incompatible length.

**Comment by markshannon** [comment]
So the bug is in the PEP, not the implementation?

**Comment by brandtbucher** [comment]
> So the bug is in the PEP, not the implementation?

It's not a bug, it's part of the spec. We even mention it twice:

> During failed pattern matches, some subpatterns may succeed. For example, while matching the pattern (0, x, 1) with the value [0, 1, 2], the subpattern x may succeed if the list elements are matched from left to right. The implementation may choose to either make persistent bindings for those partial matches or not. User code including a match statement should not rely on the bindings being made for a failed match, but also shouldn't assume that variables are unchanged by a failed match. This part of the behavior is left intentionally unspecified so different implementations can add optimizations, and to prevent introducing semantic restrictions that could limit the extensibility of this feature.

File: Lib/test/libregrtest/pgo.py
Comment (brandtbucher) [inline Lib/test/libregrtest/pgo.py:38]: FYI, I'm on the fence about this... I only added it because we now have a bunch of new code in the interpreter core with zero PGO coverage.

I might run the pyperformance suite tonight with and without `test_patma` included here, and just revert this if there are any performance regressions.

File: Python/compile.c
Comment (markshannon) [reply Python/compile.c]: Could you give this a more meaningful name, please. E.g.  `RETURN_IF_FALSE`

**Comment by asottile** [comment]
one thing I noticed (not sure if it's intentional or an oversight): `Ellipsis` / `...` are not matched as literals.  should those be included (similar to `True `/ `False` / `None`)?

**Comment by brandtbucher** [comment]
@asottile I believe we briefly considered this but decided against it. If I remember correctly, it was to avoid confusion (since it looks a bit like a wildcard-y construct):

```
match items:
    case [first, ..., last]: pass
```

If you're glancing over that code, you could be forgiven for thinking it does something like this:

```
match items:
    case [first, *_, last]: pass
```

There may have also been some worry about it appearing to be some sort of range match:

```
match character:
    case "A", ..., "Z": pass
    case "a", ..., "z": pass
```

(Note that `...` is currently disallowed in the pattern grammar, so we could always allow it later if a compelling use-case were raised.)

**Comment by brandtbucher** [comment]
(Also `Ellipsis` is just a builtin name, not a literal like `True`/`False`/`None`.)

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c:3956]: At the top-level, this bytecode does three things, and its behavior depends on `oparg`. It should be broken up, and the test moved to the compiler. I would name the first part `MATCH_KEYS`. The second part could be called `COLLECT_ITEMS_TO_DICT`. The third part is simply `LOAD_CONST True`.

The parts could do with breaking down further, but that can wait.

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: Could you make `match_keys` return `False` if it fails without error?

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: Could you replace this, and other NULL checks, with `values == NULL`

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: This bytecode seems unnecessary.
What is wrong with?
```
ROT_TWO
LOAD_CONST  oparg+1
BINARY_SUBTRACT
BINARY_SUBSCRIPT
```

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: This can be done with pre-existing instructions.

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: Why can't you use slicing, once you've done the bounds check?

File: Python/ceval.c
Comment (markshannon) [inline Python/ceval.c]: This comment basically describes the bytecode sequence that should be generated. Why not do this in the compiler?

**Review by markshannon — CHANGES_REQUESTED** [review]
There is a lot work being done in the interpreter and helper functions that should be done in the compiler.

The `MATCH_KEYS` bytecode should be split up.
`GET_INDEX` and `GET_INDEX_SLICE` can be eliminated.

The `MATCH_CLASS` and `MATCH_MAPPING` bytecodes also need breaking up, but that could wait.

File: Python/compile.c
Comment (markshannon) [inline Python/compile.c]: How does this differ from `compiler_visit_expr1(c, p)` when passed an attribute expression?
I suspect that you could replace all uses of this with `compiler_visit_expr1(c, p)`.

File: Python/compile.c
Comment (markshannon) [inline Python/compile.c]: Replace uses of this function with `ADDOP_LOAD_CONST(c, p->v.Constant.value);`
The asserts should have been checked in the verification step. Aren't they?

File: Python/compile.c
Comment (markshannon) [inline Python/compile.c:5623]: This is more an observation than a request:
The test for irrefutability (is that a real word?) seem to be passed down the stack.
Would passing it as an argument make sense?

**Comment by brandtbucher** [comment]
Thanks! I'll try to address most of these today.

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: I don't think slicing (like negative indexing) is actually part of the `collections.abc.Sequence` protocol. `deque` doesn't support it, for example:

```py
>>> import collections
>>> dq = collections.deque(range(5))
>>> isinstance(dq, collections.abc.Sequence)
True
>>> dq[1:4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sequence index must be integer, not 'slice'
```

I think it's reasonable to expect that a lot of Sequences out in the wild always expect `__getitem__` to pass a nonnegative integer index. I still want those classes to work here.

Comment (markshannon) [reply Python/ceval.c]: OK. That makes sense.
What about `UNPACK_EX`? That seems perfect for a sequence pattern with a star sub-pattern.

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: I originally added the `GET_INDEX` and `GET_INDEX_END` opcodes to try to avoid unnecessary conversions between `PyLongObject*` and `Py_ssize_t`. As you point out, though, it's probably a nano-optimization that's not worth the maintenance burden.

I think I agree that `BINARY_SUBCR` is a fine replacement for both right now.

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: Well, because this isn't *actually* what we're doing here... it's just approximately equivalent Python code to help explain the behavior, which can be a bit subtle. Call it pseudocode, if you want.

Setting up an exception handler and actually performing all of this logic in Pythonland is needlessly complicated for what we're trying to do.

File: Python/compile.c
Comment (brandtbucher) [reply Python/compile.c:5623]: Everything in `pattern_context` originally started as explicit arguments. Eventually I got tired of keeping track of everything, and just threw all of that state into the struct.

Personally, I find it easier to reason about all of these routines if they have the same signature, stack effect, etc. So I'm slightly in favor of keeping it this way... especially since I anticipate that `pattern_context` will grow more members in the future (to facilitate more efficient matching strategies, provide better warnings/errors, etc.).

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: > What about `UNPACK_EX`? That seems perfect for a sequence pattern with a star sub-pattern.

Yeah, I agree!

So (brainstorming here), we'll have two paths for compiling sequence patterns:
- One that uses `UNPACK_EX`, for cases where we have a non-wildcard star pattern.
- One that uses `BINARY_SUBSCR`, for everything else.

That way, patterns like `[first, *_, last]` stay efficient.

**Comment by brandtbucher** [comment]
Okay, I've broken off `COPY_DICT_WITHOUT_KEYS` from `MATCH_KEYS`. I've also gotten rid of all of the `GET_INDEX*` instructions.

`compiler_pattern_sequence` is a mess right now, but it works. I'm going to refactor and clean it up tomorrow, so I don't recommend burning a lot of time on it.

File: Python/ceval.c
Comment (markshannon) [reply Python/ceval.c]: It sounds like you are saying that this comment is misleading. Please remove it, or fix it.

File: Python/ceval.c
Comment (markshannon) [reply Python/ceval.c]: I'd be inclined to use `UNPACK_EX` for `[first, *_, last]` as well.
I suspect that `[first, *_, last]` is a rare pattern, and the cost of the temporary list won't be that great.

**Comment by brandtbucher** [comment]
Update: I saw an opportunity to greatly improve the emitted bytecode, so I did a slight refactoring of some of the compiler routines today.

The main change is that we no longer preserve the subject of the match as a convention in `compiler_pattern_*` routines... now they just consume it from the stack, replacing it with `True` or `False`. This avoids lots of wasteful `DUP_TOP`s, and simplifies the control flow to a level that the peephole/CFG passes can infer and optimize much more effectively. You can see commit 3f8fdaf for full diff.

Here's a before-and-after for a simple snippet:

```py
match x:
    case a, b:
        pass
```

```py
  1           0 LOAD_NAME                0 (x)

  2           2 MATCH_SEQUENCE
              4 POP_JUMP_IF_FALSE       58
              6 GET_LEN
              8 LOAD_CONST               0 (2)
             10 COMPARE_OP               2 (==)
             12 POP_JUMP_IF_FALSE       58
             14 DUP_TOP
             16 UNPACK_SEQUENCE          2
             18 DUP_TOP
             20 STORE_NAME               1 (a)
             22 LOAD_CONST               1 (True)
             24 ROT_TWO
             26 POP_TOP
             28 POP_JUMP_IF_FALSE       46
             30 DUP_TOP
             32 STORE_NAME               2 (b)
             34 LOAD_CONST               1 (True)
             36 ROT_TWO
             38 POP_TOP
             40 POP_JUMP_IF_FALSE       48
             42 LOAD_CONST               1 (True)
             44 JUMP_FORWARD             4 (to 50)
        >>   46 POP_TOP
        >>   48 LOAD_CONST               2 (False)
        >>   50 POP_JUMP_IF_FALSE       58
             52 POP_TOP
             54 LOAD_CONST               3 (None)
             56 RETURN_VALUE
        >>   58 POP_TOP
             60 LOAD_CONST               3 (None)
             62 RETURN_VALUE
```

```py
  1           0 LOAD_NAME                0 (x)

  2           2 MATCH_SEQUENCE
              4 POP_JUMP_IF_FALSE       24
              6 GET_LEN
              8 LOAD_CONST               0 (2)
             10 COMPARE_OP               2 (==)
             12 POP_JUMP_IF_FALSE       24
             14 UNPACK_SEQUENCE          2
             16 STORE_NAME               1 (a)
             18 STORE_NAME               2 (b)
             20 LOAD_CONST               1 (True)
             22 JUMP_FORWARD             4 (to 28)
        >>   24 POP_TOP
             26 LOAD_CONST               2 (False)
        >>   28 POP_JUMP_IF_FALSE       34
             30 LOAD_CONST               3 (None)
             32 RETURN_VALUE
        >>   34 LOAD_CONST               3 (None)
             36 RETURN_VALUE
```

**Comment by brandtbucher** [comment]
I'd encourage anybody who hasn't already begun reviewing this PR to start soon! We have only two more alphas, and the next is in just 10 days.

@pablogsal, when's the latest you'd be comfortable merging this before the release? I assume by the end of next week?

**Comment by pablogsal** [comment]
> when's the latest you'd be comfortable merging this before the release?

Do you mean before the alpha release? Given that a lot of people would be waiting for the next alpha to play with this, we can merge it even on the same day if is required. We can stay in contact via email if it comes to that. Or are you referring to beta freeze?

In any case, I would like to make a review pass, but these weeks I have been fighting lots of stuff and it has been impossible to find some time for it. Hopefully, I can get to it at the weekend :)

**Comment by brandtbucher** [comment]
Gotcha. I thought maybe you needed a day or two between tagging and releasing.

**Comment by pablogsal** [comment]
> I thought maybe you might need a day or two between tagging and releasing or something.

No, although is a multi-hour process, I normally do it on the same day as the release, at least for the alphas.

**Review by willingc — COMMENTED** [review]
@brandtbucher I did an initial pass and it looks good for alpha. I will look at it more closely this weekend.

**Comment by asottile** [comment]
found a bit of an odd case while trying this out -- not sure if it's intentional:

```python
x = 1

match x:
    case 1 as y if 0:
        print('unreachable')
print(f'{y=}')  # expect NameError
```

instead of NameError I get `y=1`

**Comment by pablogsal** [comment]
> found a bit of an odd case while trying this out -- not sure if it's intentional:
> 
> ```python
> x = 1
> 
> match x:
>     case 1 as y if 0:
>         print('unreachable')
> print(f'{y=}')  # expect NameError
> ```
> 
> instead of NameError I get `y=1`

IIRC I think that is covered in https://www.python.org/dev/peps/pep-0634/#side-effects-and-undefined-behavior, although the peephole could detect the branch and optimize it away.

**Comment by brandtbucher** [comment]
Yeah, we're required to bind any names used by the guard before evaluating it (obviously). The current naive implementation accomplishes by binding all names before hitting *any* guard. Strange example, but legal by the PEP.

I'm actually not sure the peephole/CFG could catch this (it can optimize out the body of the case, though, and it actually does here). Might be something for the AST optimizer.

File: Python/ceval.c
Comment (pablogsal) [inline Python/ceval.c]: This is dangerous. You are preparing a tuple with `nkeys` but then you are filling it calling `PyObject_CallFunctionObjArgs`, which can execute arbitrary Python code. At this point, the tuple is tracked by the GC so code executed by `PyObject_CallFunctionObjArgs` could see the tuple half-initialized (the same if there is a gc pass). The same thing applies to anything done meanwhile the tuple is not ready that can call into arbitrary Python.

File: Python/ceval.c
Comment (pablogsal) [inline Python/ceval.c]: Same as https://github.com/python/cpython/pull/22917/files?file-filters%5B%5D=.asdl&file-filters%5B%5D=.c&file-filters%5B%5D=.h&file-filters%5B%5D=.py#r579678887

File: Python/ceval.c
Comment (brandtbucher) [reply Python/ceval.c]: ~I see. So is filling it with `None` the right thing to do here, or can we cheat and just untrack/retrack?~

Comment (brandtbucher) [reply Python/ceval.c]: Actually, filling it with `None` can lead to problems (as we've seen before). So should we just untrack it temporarily?

I guess we could just fill a list and convert it to a tuple, too.

Comment (brandtbucher) [reply Python/ceval.c]: Okay, I just went with list -> tuple for both. Let me know if you prefer something else.

Comment (pablogsal) [reply Python/ceval.c]: Another option it to untrack the tuple after its creation and track it back when is ready. Be careful because some APIs can retrack it (like resizing the tuple). I would recommend to add some asserts to check that is untracked where is supposed to be if you decide to go this way.

File: Python/ceval.c
Comment (pablogsal) [inline Python/ceval.c]: https://github.com/python/cpython/pull/24598

:wink:

**Comment by brandtbucher** [comment]
I believe I have addressed all of the comments so far. If anyone still has additional feedback (or needs the weekend to review), please let me know. Otherwise, I'll plan on merging this Friday.

@pablogsal, should we hit it with the buildbots again?

**Comment by brandtbucher** [comment]
The three failed jobs look unrelated to these changes; I see the same failures (with the same logs) on #24643.

The 7 "pending" jobs appear to have lost connections. I'm not too familiar with the buildbots yet... is it sufficient to just press the "Rebuild" button at the top of the results page?

**Comment by brandtbucher** [comment]
Okay, I'm going to merge this. 🍾

Big thanks to everyone who took time to help out here!

File: Python/ceval.c
Comment (serhiy-storchaka) [inline Python/ceval.c:985]: It is better to use `_PyObject_LookupAttr()` here. See #106303.

File: Python/ceval.c
Comment (serhiy-storchaka) [inline Python/ceval.c:1038]: Same as above.

**Pull request merged**
